raclients/src/frontend/mod.rs

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(all(not(feature = "std"), feature = "alloc"))]
+extern crate alloc;
+
+pub(crate) mod lib {
+    pub mod core {
+        #[cfg(all(feature = "std", not(feature = "alloc")))]
+        pub use std::*;
+
+        #[cfg(all(not(feature = "std"), feature = "alloc"))]
+        pub use core::*;
+    }
+
+    pub mod alloc {
+        #[cfg(all(feature = "std", not(feature = "alloc")))]
+        pub use std::*;
+
+        #[cfg(all(not(feature = "std"), feature = "alloc"))]
+        pub use super::super::alloc::*;
+    }
+
+    pub use self::alloc::{
+        format,
+        string::{String, ToString},
+        vec::Vec,
+    };
+}
+
+#[cfg(all(feature = "std", not(feature = "alloc")))]
+pub mod unix;
+
+mod proxy;
+
+pub use proxy::*;
+
+use lib::*;
+
+use serde::{Deserialize, Serialize};
+use serde_json::{Map, Value};
+
+#[derive(Deserialize, Serialize)]
+pub enum NegotiationParam {
+    RsaPubkeyN,
+    RsaPubkeyE,
+    Bytes(Vec<u8>),
+}
```

Comment:
Please take a look at the implementation at https://github.com/stefano-garzarella/RAClients/blob/708ff568c15de211f10e58a007adedd253f38c44/src/in_proxy/clients.rs#L27

Why you discarded it?

Reply:
It's not that I discarded it, but changed it a bit. Standard procedure is to hash the public key N, public key E, and nonce from server in REPORT_DATA.

But in the original NegotiationParam, what is the order in which a Key is hashed? ordering matters, and Key does not guarantee the N and E will be hased in order. I thought this more cleary defines the order in which things should be hashed.

Also, I think's its better to hash bytes directly instead of strings, and strings need to be converted to bytes at some point anyway.

Reply:
> It's not that I discarded it, but changed it a bit. Standard procedure is to hash the public key N, public key E, and nonce from server in REPORT_DATA.

Okay, but please keep the documentation.

> 
> But in the original NegotiationParam, what is the order in which a Key is hashed? ordering matters, and Key does not guarantee the N and E will be hased in order. I thought this more cleary defines the order in which things should be hashed.

Okay.

> 
> Also, I think's its better to hash bytes directly instead of strings, and strings need to be converted to bytes at some point anyway.

If we will use JSON, we need to define an encoding for bytes (e.g. hex, Base64, etc.), for that reason a string is better. KBS, virtee do the same IIRC.